### DISCUSSION TITLE
Software idea: Play-along and looping

### DISCUSSION THREAD

**Discussion by nh2** [cardonabits/haxo-hw#41] [Ideas]
I just watched your update

https://www.crowdsupply.com/cardona-bits/haxophone/updates/manufacturing-plans-and-more

where you `aplay` sound via the serial port's console.

I think a cool feature would be that you can load some music onto your Haxophone, and then you can use it for play-along!

* Key combination for loading/switching the music file (text-to-speech could announce which one you're on)
* Normal music control key combinations: Start/stop, playback speed, back-by-10-seconds so that one can practice specific parts repeatedly
* More Haxophone-specific key combinations: Transpose (so it's easier to play of the music is in another key), either by transposing the music or by changing the Haxophone's pitch to match the music

More involved ideas:

* Turning the Haxophone into a loop station, making it possible to record and control recorded loops with key combinations, to become a single-person orchestra!
  This might make the Haxophone quite attractive even for more professional performers.

**Comment by jcard0na** [comment]
Those are good points to kick off this design.  At the moment I have implemented a special combination (draw + Bb fingering) to enter "control mode", currently only used [to switch instruments](https://github.com/cardonabits/haxo-hw/blob/main/docs/instructions_safety.md#changing-instrument-sounds).  It is relatively easy to extend to other fingerings to trigger other commands. 

Probably the biggest challenge is how to transfer the music over to the haxophone.  The serial console is not well suited for that.  Networking/wifi would probably be better.  Or even a special partition on the SD card: one could insert the SD card on a laptop, copy over the backing tracks onto a dedicated read/write partition that would then be accessible from the haxophone.

**Comment by thuntjazz** [comment]
I would love to see a transposing function!  I like that the default is in Bb, just like a tenor or soprano; that's how I have my other wind synths set up by default.  But it would also be really nice to be able to switch to, say, Eb or C really easily.

**Reply by jcard0na** [reply]
Makes sense.  Captured the feature request [here](https://github.com/cardonabits/haxo-rs/issues/19)
